File: Python/234_palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return True
        fast = head
        second = head
        while (fast.next is not None and fast.next.next is not None ):
            fast = fast.next.next
            second = second.next
        
        second_half = self.reverse(second.next)
        first_half = head

        while (second_half is not None and first_half is not None):
            if second_half.val != first_half.val:
                return False
            second_half= second_half.next
            first_half = first_half.next
        return True

    def reverse(self, second_half:ListNode):
        prev = None
        cur = second_half
        while cur is not None:
            temp = cur.next
            cur.next = prev
            prev = cur
            cur = temp
        return prev

File: Java/234_Palindrome_Linked_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// Time O(N)
// Space O(1)

class Solution {
    public boolean isPalindrome(ListNode head) {
        if(head == null) return true;
        ListNode firstHalfEnd = endOfFirstHalf(head);
        ListNode secondHalfStart = reverseList(firstHalfEnd.next);
        
        ListNode p1 = head, p2 = secondHalfStart;
        boolean result = true;
        while(result && p2 != null ){
            if(p1.val != p2.val) result = false;
            p1 = p1.next;
            p2 = p2.next;
        }
        firstHalfEnd.next = reverseList(secondHalfStart);
        return result;
    }
    private ListNode reverseList(ListNode head) {
        ListNode prev = null, curr = head;
        while(curr != null){
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;            
        }
        return prev;
    }
    private ListNode endOfFirstHalf(ListNode head) {
        ListNode slow = head, fast = head;
        while(fast.next != null && fast.next.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    }
}
